src/server/services/wms/qgswmsrenderer.cpp

```diff
@@ -644,131 +644,161 @@ namespace QgsWms
     {
-      QgsMessageLog::logMessage( QStringLiteral( "%1 // %2" ).arg( it.key(), it.value() ) );
+      throw QgsBadRequestException( QStringLiteral( "ParameterMissing" ),
+                                    QStringLiteral( "I/J parameters are required for GetFeatureInfo" ) );
     }
 
-    readLayersAndStyles( mParameters, layersList, stylesList );
-    initializeSLDParser( layersList, stylesList );
+    // get layers parameters
+    QList<QgsMapLayer *> layers;
+    QList<QgsWmsParametersLayer> params = mWmsParameters.layersParameters();
+
+    // init layer restorer before doing anything
+    std::unique_ptr<QgsLayerRestorer> restorer;
+    restorer.reset( new QgsLayerRestorer( mNicknameLayers.values() ) );
+
+    // init stylized layers according to LAYERS/STYLES or SLD
+    QString sld = mWmsParameters.sld();
+    if ( !sld.isEmpty() )
+      layers = sldStylizedLayers( sld );
+    else
+      layers = stylizedLayers( params );
 
+    // create the mapSettings and the output image
     QgsMapSettings mapSettings;
     std::unique_ptr<QImage> outputImage( createImage() );
 
+    // configure map settings (background, DPI, ...)
     configureMapSettings( outputImage.get(), mapSettings );
 
+    QgsMessageLog::logMessage( "mapSettings.destinationCrs(): " +  mapSettings.destinationCrs().authid() );
     QgsMessageLog::logMessage( "mapSettings.extent(): " +  mapSettings.extent().toString() );
     QgsMessageLog::logMessage( QStringLiteral( "mapSettings width = %1 height = %2" ).arg( mapSettings.outputSize().width() ).arg( mapSettings.outputSize().height() ) );
     QgsMessageLog::logMessage( QStringLiteral( "mapSettings.mapUnitsPerPixel() = %1" ).arg( mapSettings.mapUnitsPerPixel() ) );
 
-    //find out the current scale denominator and set it to the SLD parser
     QgsScaleCalculator scaleCalc( ( outputImage->logicalDpiX() + outputImage->logicalDpiY() ) / 2, mapSettings.destinationCrs().mapUnits() );
     QgsRectangle mapExtent = mapSettings.extent();
     double scaleDenominator = scaleCalc.calculate( mapExtent, outputImage->width() );
-    mConfigParser->setScaleDenominator( scaleDenominator );
 
-    //read FEATURE_COUNT
-    int featureCount = 1;
-    if ( mParameters.contains( QStringLiteral( "FEATURE_COUNT" ) ) )
+    // remove unwanted layers (restricted layers, ...)
+    removeUnwantedLayers( layers, scaleDenominator );
+    // remove non identifiable layers
+    QStringList nonIdentifiableLayers = mProject->nonIdentifiableLayers();
+    if ( !nonIdentifiableLayers.isEmpty() )
     {
-      featureCount = mParameters[ QStringLiteral( "FEATURE_COUNT" )].toInt( &conversionSuccess );
-      if ( !conversionSuccess )
+      QList<QgsMapLayer *> wantedLayers;
+
+      Q_FOREACH ( QgsMapLayer *layer, layers )
       {
-        featureCount = 1;
+        if ( nonIdentifiableLayers.contains( layer->id() ) )
+          continue;
+
+        wantedLayers.append( layer );
```

Comment:
Removing non identifiable layers may be done in *removeUnwantedLayers* method thanks to an optional parameter indicating if we want to keep them (or not).

What do you think?